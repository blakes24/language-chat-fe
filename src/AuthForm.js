import React from "react";
import { useFormik } from "formik";
import { useHistory } from "react-router-dom";
import TextField from "@material-ui/core/TextField";
import { makeStyles } from "@material-ui/core/styles";
import Button from "@material-ui/core/Button";

const validate = (values) => {
  const errors = {};
  if (!values.email) {
    errors.email = "Required";
  } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)) {
    errors.email = "Invalid email address";
  }
  if (!values.password) {
    errors.password = "Required";
  }
  return errors;
};

const useStyles = makeStyles((theme) => ({
  root: {
    display: "flex",
    flexWrap: "wrap",
  },
  textField: {
    marginLeft: theme.spacing(1),
    marginRight: theme.spacing(1),
    width: "25ch",
  },
  divider: {
    display: "flex",
    flexDirection: "row",
    "&::before": {
      content: '""',
      flex: "1 1",
      borderBottom: "1px solid lightgray",
      margin: "auto",
    },
    "&::after": {
      content: '""',
      flex: "1 1",
      borderBottom: "1px solid lightgray",
      margin: "auto",
    },
  },
  google: {
    background: "#db4235",
    color: "white",
    "&:hover": {
      background: "#b04238",
    },
  },
  icon: {
    marginRight: "1rem"
  },
}));

function AuthForm({ text, handleSubmit, handleGoogle, handleFacebook }) {
  const history = useHistory();
  const formik = useFormik({
    initialValues: {
      email: "",
      password: "",
    },
    validate,
    onSubmit: async (values) => {
      try {
        await handleSubmit(values);
        history.push("/");
      } catch (err) {
        formik.errors.password = err;
      }
    },
  });

  const classes = useStyles();

  return (
    <>
      <form
        className={classes.root}
        autoComplete="off"
        onSubmit={formik.handleSubmit}
      >
        <TextField
          id="email"
          name="email"
          label="Email"
          variant="outlined"
          style={{ margin: 8 }}
          InputProps={{ disableUnderline: true }}
          size="small"
          fullWidth
          onChange={formik.handleChange}
          onBlur={formik.handleBlur}
          value={formik.values.email}
          helperText={
            formik.errors.email && formik.touched.email && formik.errors.email
          }
        />
        <TextField
          id="password"
          name="password"
          label="Password"
          type="password"
          autoComplete="current-password"
          variant="outlined"
          style={{ margin: 8 }}
          InputProps={{ disableUnderline: true }}
          size="small"
          fullWidth
          onChange={formik.handleChange}
          onBlur={formik.handleBlur}
          value={formik.values.password}
          helperText={
            formik.errors.password &&
            formik.touched.password &&
            formik.errors.password
          }
        />
        <Button
          variant="contained"
          color="secondary"
          type="submit"
          fullWidth
          size="large"
          style={{ margin: 8 }}
        >
          {text}
        </Button>
      </form>
      <h4 className={classes.divider}>OR</h4>
      <div className={classes.root}>
        <Button
          variant="contained"
          type="submit"
          fullWidth
          size="large"
          style={{ margin: 8 }}
          className={classes.google}
          onClick={handleGoogle}
        >
          {`${text} with Google`}
        </Button>
        <Button
          variant="contained"
          color="secondary"
          type="submit"
          fullWidth
          size="large"
          style={{ margin: 8 }}
          onClick={handleFacebook}
        >
          {` ${text} with Facebook`}
        </Button>
      </div>
    </>
  );
}

export default AuthForm;
