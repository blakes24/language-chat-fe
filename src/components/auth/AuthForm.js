import React, { useState } from "react";
import { useFormik } from "formik";
import TextField from "@material-ui/core/TextField";
import Button from "@material-ui/core/Button";
import SocialButton from "./SocialButton";
import { faGoogle } from "@fortawesome/free-brands-svg-icons";
import FacebookIcon from "@material-ui/icons/Facebook";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import Loading from "../Loading";
import { useStyles } from "./AuthFormStyles";

const validate = (values) => {
  const errors = {};
  if (!values.email) {
    errors.email = "Required";
  } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.email)) {
    errors.email = "Invalid email address";
  }
  if (!values.password) {
    errors.password = "Required";
  } else if (values.password.includes(" ")) {
    errors.password = "Invalid password (cannot contain spaces)";
  }
  if (values.passwordConfirm) {
    if (values.password !== values.passwordConfirm) {
      errors.passwordConfirm = "Passwords do not match";
    }
  }
  return errors;
};

function AuthForm({
  text,
  handleSubmit,
  handleGoogle,
  handleFacebook,
  handleFail,
  newUser,
}) {
  const [loading, setLoading] = useState(false);

  const formik = useFormik({
    initialValues: {
      email: "",
      password: "",
      passwordConfirm: "",
    },
    validate,
    onSubmit: async (values) => {
      setLoading(true);
      try {
        await handleSubmit(values);
      } catch (err) {
        setLoading(false);
        newUser
          ? (formik.errors.passwordConfirm = err)
          : (formik.errors.password = err);
      }
    },
  });

  const classes = useStyles();

  return (
    <>
      {loading && <Loading />}
      <form
        className={classes.root}
        autoComplete="off"
        onSubmit={formik.handleSubmit}
      >
        <TextField
          id="email"
          name="email"
          label="Email"
          variant="outlined"
          style={{ margin: 8 }}
          size="small"
          fullWidth
          onChange={formik.handleChange}
          onBlur={formik.handleBlur}
          value={formik.values.email}
          FormHelperTextProps={{ error: true }}
          helperText={
            formik.errors.email && formik.touched.email && formik.errors.email
          }
        />
        <TextField
          id="password"
          name="password"
          label="Password"
          type="password"
          autoComplete="current-password"
          variant="outlined"
          style={{ margin: 8 }}
          size="small"
          fullWidth
          onChange={formik.handleChange}
          onBlur={formik.handleBlur}
          value={formik.values.password}
          FormHelperTextProps={{ error: true }}
          helperText={
            formik.errors.password &&
            formik.touched.password &&
            formik.errors.password
          }
        />
        {newUser && (
          <TextField
            id="passwordConfirm"
            name="passwordConfirm"
            label="Confirm Password"
            type="password"
            autoComplete="current-password"
            variant="outlined"
            style={{ margin: 8 }}
            size="small"
            fullWidth
            onChange={formik.handleChange}
            onBlur={formik.handleBlur}
            value={formik.values.passwordConfirm}
            FormHelperTextProps={{ error: true }}
            helperText={
              formik.errors.passwordConfirm &&
              formik.touched.passwordConfirm &&
              formik.errors.passwordConfirm
            }
          />
        )}
        <Button
          variant="contained"
          color="secondary"
          type="submit"
          fullWidth
          size="large"
          style={{ margin: 8 }}
        >
          {text}
        </Button>
      </form>
      <h4 className={classes.divider}>OR</h4>
      <div className={classes.root}>
        <SocialButton
          color="secondary"
          className={classes.facebook}
          provider="facebook"
          appId={process.env.REACT_APP_FACEBOOK_APP_ID}
          onLoginSuccess={handleFacebook}
          onLoginFailure={handleFail}
        >
          <FacebookIcon className={classes.icon} />
          {` ${text} with Facebook`}
        </SocialButton>
        <SocialButton
          color="secondary"
          className={classes.google}
          provider="google"
          appId={process.env.REACT_APP_GOOGLE_CLIENT_ID}
          onLoginSuccess={handleGoogle}
          onLoginFailure={handleFail}
        >
          <FontAwesomeIcon className={classes.icon} icon={faGoogle} />{" "}
          {`${text} with Google`}
        </SocialButton>
      </div>
    </>
  );
}

export default AuthForm;
