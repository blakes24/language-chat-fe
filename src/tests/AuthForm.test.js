import { render } from "@testing-library/react";
import AuthForm from "../components/auth/AuthForm";
import { MemoryRouter } from "react-router";

test("renders without crashing", () => {
  render(
    <MemoryRouter>
      <AuthForm />
    </MemoryRouter>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(<AuthForm />);
  expect(asFragment()).toMatchSnapshot();
});
