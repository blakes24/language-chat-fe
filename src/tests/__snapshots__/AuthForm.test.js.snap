// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form
    autocomplete="off"
    class="makeStyles-root-11"
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      style="margin: 8px;"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
        data-shrink="false"
        for="email"
        id="email-label"
      >
        Email
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
          id="email"
          name="email"
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-17 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-19"
          >
            <span>
              Email
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      style="margin: 8px;"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
        data-shrink="false"
        for="password"
        id="password-label"
      >
        Password
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
      >
        <input
          aria-invalid="false"
          autocomplete="current-password"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
          id="password"
          name="password"
          type="password"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-17 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-19"
          >
            <span>
              Password
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-containedSizeLarge MuiButton-sizeLarge MuiButton-fullWidth"
      style="margin: 8px;"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiButton-label"
      />
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </form>
  <h4
    class="makeStyles-divider-13"
  >
    OR
  </h4>
  <div
    class="makeStyles-root-11"
  />
</DocumentFragment>
`;
